typing/typecore.ml

```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
       approx_type env sty
   | _ -> newvar ()
 
-let type_pattern_approx env spat =
+let type_pattern_approx env spat ty_expected =
   match spat.ppat_desc with
-  | Ppat_constraint (_, sty) -> approx_type env sty
-  | _ -> newvar ()
+  | Ppat_constraint (_, sty) ->
+      let inferred_ty = approx_type env sty in
+      begin try unify_pat_types spat.ppat_loc env inferred_ty ty_expected
+        with Unify trace ->
+        raise(Error(spat.ppat_loc, env, Pattern_type_clash(trace, None)))
+      end;
+  | _ -> ()
 
-let type_approx_fun env label default spat ret_ty =
-  let ty = type_pattern_approx env spat in
-  let ty =
-    match label, default with
-    | (Nolabel | Labelled _), _ -> ty
-    | Optional _, None ->
-       unify_pat_types spat.ppat_loc env ty (type_option (newvar ()));
-       ty
-    | Optional _, Some _ ->
-       type_option ty
+let type_approx_fun_one_param
```

Comment:
Could be nice to add a comment explaining with `spato` is an optional pattern (ie : because we skip a match potentially containing a gadt)